### DISCUSSION TITLE
.Net 7 for windows (Not Maui)

### DISCUSSION THREAD

**Discussion by AskBojesen** [dotnet-bluetooth-le/dotnet-bluetooth-le#739] [Ideas]
Hi, we currently use the plugin in Maui Android and iOS.

I can get it working in Maui context in Windows, but when starting as wpf it fails du to WINUI dependency the
**CommunityToolkit.WinUI.Connectivity.BluetoothLEHelper** where the WINUI DispatcherQueue is not existing.

All the **Windows.Devices.Bluetooth.*** is working directly in .Net 7 - eg the
device = Windows.Devices.BluetoothFromBluetoothAddressAsync(ulong address) is working
Only the **CommunityToolkit.WinUI.Connectivity.BluetoothLEHelper** is not working in .Net 7 Wpf 

My point is that one could make a Windows specific implementation fulfilling the interfaces in **Plugin.BLE.Abstractions.Contracts**.*

I could do this implementation - what are your thoughts ?

**Comment by janusw** [comment]
So, IIUC your proposal is to make the Windows implementation more general by only relying on `Windows.Devices.Bluetooth` directly, but removing any dependencies on `CommunityToolkit.WinUI` or `Microsoft.Toolkit.Uwp` (we use the latter for the Xamarin.UWP case). Correct?

[Or are you saying you would rather add a third variant of the Windows implementation (for WPF), in addition to the two we already have?]

IMHO the idea sounds great in principle. In particular a generic implementation that works with all UI frameworks would be nice. If you could provide such an implementation (with equivalent functionality to the current one), we'd certainly be happy to accept it.

[Note that I don't use the Windows implementation and don't have enough knowledge of the Windows-specific APIs to judge if such a project is feasible and reasonable.]

**Reply by AskBojesen** [reply]
Today I finished a proof of concept. Rewriting the Windows implementation of the Plugin NOT to use `CommunityToolkit.WinUI.Connectivity.BluetoothLEHelper` and the `CommunityToolkit.WinUI.Connectivity.ObservableBluetoothLEDevice` that are only working if a WINUI `Microsoft.UI.Dispatching.DispatcherQueue` is present.
I managed to have a console application running in Windows communicating with our own BLE device with TargetFramework=net7.0-windows10.0.19041.0

My need is to use the Plugin.BLE in a Windows WPF/console application. We already use the Plugin.BLE for Android and Apple, where it works nicely. I don't know about the Xamarin.UWP - my background is Windows c# with Forms and Wpf.
Maybe my new implementation can also be used in a UWP app - I guess so. It's only the dependence to the `Microsoft.UI.Dispatching.DispatcherQueue` we need to get rid of in order to have a generic Windows implementation.

I will clean up my proof of concept and push to a branch, then you can give it a look.

**Reply by AskBojesen** [reply]
@janusw : I would like to make a branch to show my windows BLE console client demo.
I guess a branch named `feature/winconsole` based on master ?
Will you create such a branch or give me the rights?

**Reply by janusw** [reply]
Please just fork the repo and create a branch in your own fork (you have all the necessary permissions there).

**Reply by AskBojesen** [reply]
I have a working implementation with windows console demo here: https://github.com/AskBojesen/dotnet-bluetooth-le/tree/feature/winconsole

What do you think?

![image](https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/assets/131768758/61441c2e-6df5-4256-8a89-b0b0815c574c)

**Comment by AskBojesen** [comment]
Hi @DamianSuess can you see this?

I see that you are using this BLE.Plugin for UWP.

I have a suggestion for a refactoring - can you give this suggestion a look?

**Comment by AskBojesen** [comment]
Discussion is handled in: https://github.com/dotnet-bluetooth-le/dotnet-bluetooth-le/pull/746
